File: 64.最小路径和.swift
/*
 * @lc app=leetcode.cn id=64 lang=swift
 *
 * [64] 最小路径和
 */

// @lc code=start
class Solution {
    func minPathSum(_ grid: [[Int]]) -> Int {
        let rows = grid.count
        let cols = grid[0].count
        var dps = Array(repeating: Array(repeating: 0, count: cols), count: rows)
        dps[0][0] = grid[0][0]
        for idx in 1..<rows {
            dps[idx][0] = dps[idx-1][0] + grid[idx][0]
        }
        for idx in 1..<cols {
            dps[0][idx] = dps[0][idx-1] + grid[0][idx]
        }
        for index in 1..<rows {
            for jndex in 1..<cols {
                dps[index][jndex] = min(dps[index-1][jndex], dps[index][jndex-1]) + grid[index][jndex]
            }
        }
        return dps[rows-1][cols-1]
    }
}
// @lc code=end

File: 64.最小路径和.cpp
/*
 * @lc app=leetcode.cn id=64 lang=cpp
 *
 * [64] 最小路径和
 */

// @lc code=start
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int rows = grid.size(), cols = grid[0].size();
        auto dp = vector< vector<int> >(rows, vector<int>(cols));
        dp[0][0] = grid[0][0];
        for (int i = 1; i < rows; i++) {
            dp[i][0] = dp[i-1][0] + grid[i][0];
        }
        for (int j = 1; j < cols; j++) {
            dp[0][j] = dp[0][j-1] + grid[0][j];
        }
        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j];
            }
        }
        return dp[rows-1][cols-1];
    }
};
// @lc code=end

